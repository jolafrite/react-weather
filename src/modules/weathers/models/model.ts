export interface IFetchWeather extends IWeather {}

export interface IWeather {
  city: {
    id: number;
    name: string;
    coord: {
      lon: number;
      lat: number;
    };
    country: string;
  };
  cnt: number;
  list: IWeatherList[];
};

export interface IWeatherList {
  dt: number;
  main: {
    temp: number;
    temp_min: number;
    temp_max: number;
    pressure: number;
    humidity: number;
    sea_level: number;
    grnd_level: number;
    temp_kf: number;
  };
  weather: IWeatherListWeather[];
  clouds: {
    all: number;
  };
  wind: {
    speed: number;
    deg: number;
  };
  dt_txt: string;
};

export interface IWeatherListWeather {
  id: number;
  main: string;
  description: string;
  icon: string;
}

export interface IWeatherLast5DaysForecast {
  [date: string]: IWeatherDailyForecast;
}

export interface IWeatherDailyForecast {
  temp: number;
  temp_max: number;
  temp_min: number;
  hourly: {
    [time: string]: IWeatherList;
  };
}
