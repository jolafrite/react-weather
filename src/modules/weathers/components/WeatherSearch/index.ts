export * from './WeatherSearch';
export { default } from './WeatherSearch';
