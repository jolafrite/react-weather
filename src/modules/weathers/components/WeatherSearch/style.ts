import { makeStyles, createStyles } from "@material-ui/core";
import { rem } from "polished";

export const useStyles = makeStyles(theme =>
  createStyles({
    root: {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      height: '100%',
      width: '100%',
      cursor: 'default',
    },
    card: {
      width: 800,
      maxHeight: '100%',
      maxWidth: '100%',
    },
    notFound: {
      margin: 0,
      padding: rem(15),
    },
    error: {
      margin: 0,
      padding: rem(15),
      backgroundColor: theme.palette.error.main,
    }
  })
);
