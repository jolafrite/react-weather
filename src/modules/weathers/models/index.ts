export * from './weather.fixtures';
export * from './model';
