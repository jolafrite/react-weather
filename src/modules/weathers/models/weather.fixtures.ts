import { useFixtureFactory } from "../../../common/utils/test";
import { IWeather } from "./model";
import { ISearchWeather } from "../../../fetch/search-weather/model";

export const weatherFixtures = useFixtureFactory<IWeather>({
  city: {
    id: 1851632,
    name: "Shuzenji",
    coord: { lon: 138.933334, lat: 34.966671 },
    country: "JP"
  },
  cnt: 38,
  list: [
    {
      dt: 1406106000,
      main: {
        temp: 298.77,
        temp_min: 298.77,
        temp_max: 298.774,
        pressure: 1005.93,
        sea_level: 1018.18,
        grnd_level: 1005.93,
        humidity: 87,
        temp_kf: 0.26
      },
      weather: [
        { id: 804, main: "Clouds", description: "overcast clouds", icon: "04d" }
      ],
      clouds: { all: 88 },
      wind: { speed: 5.71, deg: 229.501 },
      dt_txt: "2014-07-23 09:00:00"
    }
  ]
});

export const searchWeatherFixtures = useFixtureFactory<ISearchWeather>({
  id: 2643743,
  name: "London",
  coord: { lat: 51.5085, lon: -0.1258 },
  main: {
    temp: 284.31,
    pressure: 1010,
    humidity: 87,
    temp_min: 283.15,
    temp_max: 285.93,
    sea_level: 1,
    grnd_level: 2,
    temp_kf: 3,
  },
  dt: 1560233226,
  wind: {
    speed: 1.5,
    deg: 150
  },
  sys: {
    country: "GB"
  },
  rain: {
    "1h": 3,
    "3h": 4,
  },
  snow: {
    "1h": 1,
    "3h": 2,
  },
  clouds: { all: 90 },
  weather: [
    { id: 500, main: "Rain", description: "light rain", icon: "10d" },
    { id: 701, main: "Mist", description: "mist", icon: "50d" }
  ]
});
