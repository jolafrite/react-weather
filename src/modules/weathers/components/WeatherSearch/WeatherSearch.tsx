import React, { FC, memo } from 'react';
import { IWeather } from '../../models';
import Layout from '../../../../common/components/Layout';
import { List, ListItem, Paper } from '@material-ui/core';
import ListItemWeather from '../Weather/ListItemWeather';
import WeatherSearchForm from '../WeatherSearchForm';
import { useStyles } from './style';

export interface IWeatherSearchProps {
  onSearchWeather: (query: string) => void;
  onWeatherClick: (weather: IWeather) => void;
  weathers: IWeather[];
  loaded: boolean;
  loading: boolean;
  error: boolean;
}

const WeatherSearch: FC<IWeatherSearchProps> = props => {
  const classes = useStyles();
  const { onSearchWeather, onWeatherClick, weathers, error, loaded } = props;

  const listTemplate = (
    <List
      component="div"
      aria-label="List of weathers"
    >
      {weathers.map(weather => (
        <ListItem
          key={weather.city.id}
          button
          component="a"
          onClick={() => onWeatherClick(weather)}
        >
          <ListItemWeather
            weather={weather}
          />
        </ListItem>
      ))}
    </List>
  );

  const errorTemplate = (
    <div className={classes.error}>
      <div>The search failed,</div>
      <div>please check your Internet connection and try again</div>
    </div>
  );

  const notFoundTemplate = (
    <div className={classes.notFound}>
      <div>Result not found</div>
    </div>
  );

  return (
    <Layout>
      <div className={classes.root}>
        <Paper>
          <WeatherSearchForm
            {...props}
            onSubmit={onSearchWeather}
          />

          {error && (
            errorTemplate
          )}

          {loaded && weathers.length === 0 && (
            notFoundTemplate
          )}

          {weathers.length !== 0 && (
            listTemplate
          )}
        </Paper>
      </div>
    </Layout>
  )
};

export default memo(WeatherSearch);
