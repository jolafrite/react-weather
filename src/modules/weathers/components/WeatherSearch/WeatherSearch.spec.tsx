import React from 'react';
import { render, cleanup  } from '@testing-library/react';
import WeatherSearch, { IWeatherSearchProps } from './WeatherSearch';
import { weatherFixtures } from '../../models';

jest.mock('../../../../common/components/Layout', () => ({ children }: any) => <div className="Layout">{children}</div>);
jest.mock('../WeatherSearchForm', () => () => <div className="WeatherSearchForm" />);
jest.mock('../Weather/ListItemWeather', () => () => <div className="ListItemWeather" />);

describe('component: WeatherSearch', () => {

  afterEach(cleanup);

  const createComponent = (values = {}) => {
    const props: IWeatherSearchProps = {
      onSearchWeather: () => {},
      onWeatherClick: () => {},
      weathers: [],
      loaded: false,
      loading: false,
      error: false,
      ...values,
    };

    return render(
      <WeatherSearch
        {...props}
      />
    );
  };

  describe('init', () => {
    test('should only render the form', () => {
      const component = createComponent();

      expect(
        component.container
      ).toMatchSnapshot();
    });
  });

  describe('list of weathers passed', () => {
    test('display the list', () => {
      const weathers = [
        weatherFixtures(),
        weatherFixtures({
          city: { id: 1 }
        })
      ];

      const component = createComponent({ weathers });

      expect(
        component.container
      ).toMatchSnapshot();
    });
  });

  describe('error is set to true', () => {
    test('should show the error message', () => {
      const component = createComponent({ error: true });

      expect(
        component.container
      ).toMatchSnapshot();
    })
  });

  describe('result not found', () => {
    test('should show the no result found message', () => {
      const component = createComponent({ loaded: true });

      expect(
        component.container
      ).toMatchSnapshot();
    })
  });

});
