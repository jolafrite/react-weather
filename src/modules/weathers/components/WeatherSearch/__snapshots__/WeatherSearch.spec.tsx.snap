// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: WeatherSearch error is set to true should show the error message 1`] = `
<div>
  <div
    class="Layout"
  >
    <div
      class="makeStyles-root-91"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="WeatherSearchForm"
        />
        <div
          class="makeStyles-error-94"
        >
          <div>
            The search failed,
          </div>
          <div>
            please check your Internet connection and try again
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`component: WeatherSearch init should only render the form 1`] = `
<div>
  <div
    class="Layout"
  >
    <div
      class="makeStyles-root-1"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="WeatherSearchForm"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`component: WeatherSearch list of weathers passed display the list 1`] = `
<div>
  <div
    class="Layout"
  >
    <div
      class="makeStyles-root-32"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="WeatherSearchForm"
        />
        <div
          aria-label="List of weathers"
          class="MuiList-root MuiList-padding"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="ListItemWeather"
            />
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="ListItemWeather"
            />
            <span
              class="MuiTouchRipple-root"
            />
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`component: WeatherSearch result not found should show the no result found message 1`] = `
<div>
  <div
    class="Layout"
  >
    <div
      class="makeStyles-root-122"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="WeatherSearchForm"
        />
        <div
          class="makeStyles-notFound-124"
        >
          <div>
            Result not found
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
